File: leetcode/maychallenge/moveZeroes.go
package main

import "fmt"

func main() {
	//nums := []int{0, 1, 0, 3, 12}
	nums := []int{4, 2, 4, 0, 0, 3, 0, 5, 1, 0}
	moveZeroes(nums)
	fmt.Println(nums)
}

func moveZeroes(nums []int) {
	// for i1, i2 := 0, 0; i1 < len(nums); i1++ {
	// 	if nums[i1] == 0 {
	// 		nums[i1], nums[i2] = nums[i2], nums[i1]
	// 		i2++
	// 	}
	// }
	for i1 := range nums {
		for i2 := range nums {
			if nums[i1] != 0 && nums[i2] == 0 {
				nums[i2], nums[i1] = nums[i1], nums[i2]
			}
		}
	}
}

File: algo/moveZeroes.java
class Solution {
    // O(n^2)
    public void moveZeroes(int[] nums) {
        int i = 0;
        int end = nums.length;
        while (i < end) {
            if (nums[i]==0) {
                for (int j=i;j<nums.length-1; j++) {
                nums[j] = nums[j+1];
                }
                nums[nums.length-1]=0;
                end--;
            }  else {
                i++;
            }
        }
    }

    // O(n) sub-optimal
    public void moveZeroes2(int[] nums) {
        int countZero = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                countZero++;
            } else {
                nums[i-countZero] = nums[i];
            }
        }
        for (int i = nums.length-1; i > nums.length-1 -countZero; i-- ) {
            nums[i] = 0;
        }
    }

    // O(n) optimal
    public void moveZeroes3(int[] nums) {
        for (int lastNonZeroFoundAt = 0, cur = 0; cur < nums.length; cur++) {
            if (nums[cur] != 0) {
                swap(nums, lastNonZeroFoundAt++, cur);
            }
        }
    }
    public static void swap(int[] arr, int i, int j) {
        arr[i] = (arr[i] + arr[j]) - (arr[j] = arr[i]);
    }
}